### PR TITLE
[feature](nereids) Integrate existing stats framework to nereids

### PR DESCRIPTION
**Pull request by Kikyou1997**
# Proposed changes

Issue Number: noissue

## Problem Summary:

Write a new class named `StatsCalculator` to calculate stats info and set it to both group and operator.

## Checklist(Required)

1. Does it affect the original behavior: No)
2. Has unit tests been added: (No)
3. Has document been added or modified: (No)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/StatsCalculator.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+/**
+ * This class is used to get the statistics info for each group.
+ */
+public class StatsCalculator extends DefaultPlanVisitor<StatsDeriveResult, PlanContext> {
+
+    /**
+     * Gather statistic information of GroupExpression's parent group.
+     */
+    public void estimate(GroupExpression groupExpression) {
+        Plan plan = groupExpression.getPlan();
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/StatsCalculator.java:48]: groupExpression.getPlan() will return a plan with GroupPlan as children, so DefaultPlanVisitor will stop recursive visit plan when visit GroupPlan.

so you must rewrite visitGroupPlan to recursive visit children plans

Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/cost/StatsCalculator.java:48]: I think there is no need to recursively visit child plan, DeriveStatsJob will handle this recursively with a stack.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/StatsCalculator.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        PlanContext planContext = new PlanContext(groupExpression);
+        StatsDeriveResult stats = plan.accept(this, planContext);
+        groupExpression.getParent().setStatistics(planContext.getStatistics());
+        plan.setStatsDeriveResult(stats);
```
[the remark this answers is not in the document]
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/cost/StatsCalculator.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/PlanContext.java [hunk 3]
```diff
@@ -103,4 +109,34 @@ public StatsDeriveResult getChildStatistics(int index) {
...
+        return plan.getStats();
+    }
+
+    @Override
+    public void setStatsDeriveResult(StatsDeriveResult result) {
+        if (groupExpression != null) {
+            groupExpression.getParent().setStatistics(result);
+            groupExpression.setStatDerived(true);
+        }
+        plan.setStats(result);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/PlanContext.java:127]:
```suggestion
        } else {
            plan.setStats(result);
        }
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/PlanContext.java [hunk 4]
```diff
@@ -36,7 +42,7 @@
  * A ExpressionHandle is attached to either {@link Plan} or {@link GroupExpression}.
  * Inspired by GPORCA-CExpressionHandle.
  */
-public class PlanContext {
+public class PlanContext implements PlanStats {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/PlanContext.java:45]: maybe we need remove PlanContext and let GroupExpression implements PlanStats

**Pull request closed without merging**